### PR TITLE
Fx files finding: CMIP6: look in experiment=`piControl` as a last measure if fx data is not found

### PR DESCRIPTION
**Pull request by valeriupredoi**
## Description

This is a sort of a lifeboat approach: many CMIP6 models don't have fx data in any other experiment BUT `piControl` - fx data is produced only once and is experiment-agnostic, so they don't bother producing multiple copies; also the ESGF data storing rules are a bit lax here so they don't ask the uploaders to provide multiple copies of fx data for multiple experiments. So if one wants `exp=historical, var: tas` and then asks for some fx var to go with it, ESMValTool will not find that var, hence the lifeboat search in `piControl` right before exiting and declaring fx data not found. This **helps** with #1314 but it don't close it by no means.

Link to documentation: Q: do you think we should add documentation for this?

***

## [Checklist](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#checklist-for-pull-requests)

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [x] [🧪][2] The new functionality is [relevant and scientifically sound](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#scientific-relevance)
- [x] [🛠][1] This pull request has a [descriptive title and labels](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-title-and-label)
- [x] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#code-quality)
- [ ] [🧪][2] and [🛠][1] [Documentation](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#documentation) is available
- [x] [🛠][1] [Unit tests](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#tests) have been added
- [x] [🛠][1] Changes are [backward compatible](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#backward-compatibility)
- [x] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-checks) were successful

***

### CODE REVIEW

**Comment by valeriupredoi** [comment]
you guys think we should have a one-liner in the docs describing this feature? :beer:

File: esmvalcore/_recipe.py [hunk 1]
```diff
@@ -433,9 +433,25 @@ def _get_fx_files(variable, fx_info, config_user):
         fx_files = _get_input_files(fx_info, config_user)[0]
 
     # Flag a warning if no files are found
+    # CMIP6 special case:
+    # Before completely giving up, try looking for fx files in piControl:
+    # a lot of CMIP6 models have fx data only in piControl
     if not fx_files:
         logger.warning("Missing data for fx variable '%s'",
                        fx_info['short_name'])
+        if fx_info['project'] == 'CMIP6':
```
Comment (schlunma) [inline esmvalcore/_recipe.py:442]: I think you need to change the logic here:

```py
if not fx_files and fx_info['project'] == 'CMIP6':
    # do your thing
    # do NOT log warning if not successful
if not fx_files:
    # log the existing warning
```

With the current version, the warning about missing data is logged, even if the in the following iteration files are found for `piControl`, which is very confusing!

File: esmvalcore/_recipe.py [hunk 2]
```diff
@@ -433,9 +433,25 @@ def _get_fx_files(variable, fx_info, config_user):
...
+            logger.debug(
+                "Missing data for fx variable %s "
+                "and experiment %s; since CMIP6, "
+                "will try experiment piControl, most fx "
+                "stuff is there.", fx_info['short_name'], fx_info['exp'])
+            fx_info['exp'] = 'piControl'
```
Comment (schlunma) [inline esmvalcore/_recipe.py:448]:
```suggestion
            fx_info = dict(fx_info)
            fx_info['exp'] = 'piControl'
```

To avoid unwanted side effects.

File: esmvalcore/_recipe.py [hunk 3]
```diff
@@ -433,9 +433,25 @@ def _get_fx_files(variable, fx_info, config_user):
...
+            fx_files = _get_input_files(fx_info, config_user)[0]
+            if not fx_files:
+                logger.warning(
+                    "Failed to find data for CMIP6 fx variable "
+                    "%s with experiment "
+                    "piControl as well.", fx_info['short_name'])
```
Comment (schlunma) [inline esmvalcore/_recipe.py:454]:
```suggestion
```

I think this is not necessary, see comment above.

File: tests/integration/test_recipe.py [hunk 4]
```diff
@@ -2199,7 +2199,13 @@ def test_landmask_no_fx(tmp_path, patched_failing_datafinder, config_user):
...
         assert isinstance(fx_variables, dict)
         fx_variables = fx_variables.values()
-        assert not any(fx_variables)
+        # usable fx data is found for CMIP6 in piControl
+        # the test is to check if all those found instances are only CMIP6's
+        if fx_variables:
+            count_cmip6 = [
+                p for p in fx_variables if p.get('project', None) == 'CMIP6'
+            ]
+            assert len(fx_variables) == len(count_cmip6)
```
Comment (schlunma) [inline tests/integration/test_recipe.py:2208]: I'm not sure why this change is necessary. The `patched_failing_datafinder` that is used your is not dependent on the `exp` right now, so your changes here shouldn't affect the outcome of this change. Can you double-check?

File: tests/integration/test_recipe.py [hunk 5]
```diff
@@ -2397,6 +2403,73 @@ def test_fx_vars_mip_search_cmip6(tmp_path, patched_datafinder, config_user):
...
+
+    # Check add_fx_variables
+    fx_variables = product.settings['add_fx_variables']['fx_variables']
+    assert isinstance(fx_variables, dict)
+    assert len(fx_variables) == 5
+    assert '_fx_' in fx_variables['areacella']['filename']
+    assert '_Ofx_' in fx_variables['areacello']['filename']
+    assert '_Efx_' in fx_variables['clayfrac']['filename']
+    assert '_fx_' in fx_variables['sftlf']['filename']
+    assert '_Ofx_' in fx_variables['sftof']['filename']
```
Comment (schlunma) [inline tests/integration/test_recipe.py:2470]: I'm pretty sure this test is not working as it should. The `patched_datafinder` that is used here does not care about `exp`, so it also happily returns date for your `exp: historical-222`.

Please check if the files are truly from your exp by something like

```py
assert '_historical-222_' in fx_variables['areacella']['filename']
```

I'm pretty sure you need to extend these lines and use the `patched_failing_datafinder` for this to work:

https://github.com/ESMValGroup/ESMValCore/blob/671dd3799e84a0ed033e1ae66a5021eb25d60c09/tests/integration/test_recipe.py#L296-L303

**Review by schlunma — CHANGES_REQUESTED** [review]

**Comment by schlunma** [comment]
> you guys think we should have a one-liner in the docs describing this feature? beer

Yes! I think [this section](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/recipe/preprocessor.html?highlight=fx#fx-variables-as-cell-measures-or-ancillary-variables) would be a good place.

**Comment by zklaus** [comment]
> you guys think we should have a one-liner in the docs describing this feature?

I think we probably should not have this feature.

I see that you removed the "Before you get started" section of the PR template, but perhaps that does serve a purpose?
We already have means to address this problem, e.g. the wildcard specifier "*" for the experiment of an fx file. This seems to try to scratch the same itch.

I think the next step is for @ledm to clarify the problems in ESMValGroup/ESMValTool#2308.

In my view, we don't have the collective time to cook up half-solutions to ill-described problems. If you have time for ESMValTool, please work on PRs in the [2.4.0 milestone](https://github.com/ESMValGroup/ESMValCore/milestone/7) instead.

**Comment by valeriupredoi** [comment]
@zklaus 

> We already have means to address this problem, e.g. the wildcard specifier "*" for the experiment of an fx file. This seems to try to scratch the same itch.

That's very good point! But this solves two things, and it actually complements the wildcard implementation:

- this solves the problem when an exp is actually specified but fx files are not found there
- and the fact that the code should first look in the exp defined by user (if any), then piControl, and only then in `*` exps

**Comment by valeriupredoi** [comment]
> In my view, we don't have the collective time to cook up half-solutions to ill-described problems. If you have time for ESMValTool, please work on PRs in the 2.4.0 milestone instead.

The problem is very well defined in this PR's description - and is independent of us - fx data is sparse and most of the times only in piControl

File: esmvalcore/_recipe.py (hunk 1 above)
Comment (valeriupredoi) [reply esmvalcore/_recipe.py:442]: yes exactly, imo the warning needs to be there since we switch experiments, if the user defines a preferred experiment they need to be aware that fx files were not found for that one, and instead the code tried to look for piControl ones

File: esmvalcore/_recipe.py (hunk 2 above)
Comment (valeriupredoi) [reply esmvalcore/_recipe.py:448]: why do you want a copy of that, it'll get destroyed at return point anyway?

File: tests/integration/test_recipe.py (hunk 4 above)
Comment (valeriupredoi) [reply tests/integration/test_recipe.py:2208]: yes I double checked - the assertion fails for CMIP6 since it finds fx files in piControl, with it passing for CMIP5 since there were no fx files found there

File: tests/integration/test_recipe.py (hunk 5 above)
Comment (valeriupredoi) [reply tests/integration/test_recipe.py:2470]: no it does not, it returns no files for that exp for CMIP5

File: esmvalcore/_recipe.py (hunk 1 above)
Comment (schlunma) [reply esmvalcore/_recipe.py:442]: Ok, how about:


```py
if not fx_files and fx_info['project'] == 'CMIP6':
    # do your thing
    # if files found, output warning that exp has changed
if not fx_files:
    # log the existing warning
```

As I said, right now it gives a warning that no files are found, then only says in debug that `piControl` is searched now, and if actually files are found, nothing else is returned. So if your log level is not `debug`, you just get the message about missing data, nothing else (even though the `piControl` data has been found)

File: esmvalcore/_recipe.py (hunk 2 above)
Comment (valeriupredoi) [reply esmvalcore/_recipe.py:448]: ok am silly, that actually gets returned, good call!

Comment (schlunma) [reply esmvalcore/_recipe.py:449]: `dict` are mutable, so of you change that in the function it's also changed in the code the calls the function,

File: esmvalcore/_recipe.py (hunk 1 above)
Comment (valeriupredoi) [reply esmvalcore/_recipe.py:442]: true, forgot those are debug! yeah, lemme change it :+1:

File: tests/integration/test_recipe.py (hunk 4 above)
Comment (schlunma) [reply tests/integration/test_recipe.py:2208]: If this is the case then I think something weird happens in your new code or the test - the failing data finder is designed to fail in this situation, regardless of `exp`...

Comment (valeriupredoi) [reply tests/integration/test_recipe.py:2208]: the original test was `assert not any(fx_variables)` which was passing since there were no fx files for either CMIP5 or CMIP6; now, with the lookup in piControl, it finds the piControl fx file so the original test was failing for CMIP6, but still passing for CMIP5, nonetheless it was failing at exit

Comment (schlunma) [reply tests/integration/test_recipe.py:2208]: But due the use of the [`patched_failing_datafinder`](https://github.com/ESMValGroup/ESMValCore/blob/d75eb9102bf1938bb33da2034c8a25973651532d/tests/integration/test_recipe.py#L279-L306), this test shouldn't find any data, regardless of CMIP5 or CMIP6 or the `exp` of them! I'm really confused why it suddenly finds data for `piControl` now...

Comment (valeriupredoi) [reply tests/integration/test_recipe.py:2208]: gah! blast, you're right, it finds whatever you want it to find, it did find `historical-222` instead of `piControl`  :man_facepalming:

File: tests/integration/test_recipe.py (hunk 5 above)
Comment (valeriupredoi) [reply tests/integration/test_recipe.py:2470]: gah! blast, you're right, it finds whatever you want it to find, it did find historical-222 instead of piControl man_facepalming

File: tests/integration/test_recipe.py (hunk 4 above)
Comment (valeriupredoi) [reply tests/integration/test_recipe.py:2208]: it finds the piControl data all too well, man! Here's the output if I force fail the test:
```
dict_values([])
dict_values([{'preprocessor': 'landmask', 'project': 'CMIP6', 'mip': 'Lmon', 'exp': 'piControl', 'start_year': 2000, 'end_year': 2005, 'ensemble': 'r1i1p1f1', 'variable_group': 'sftof', 'short_name': 'sftof', 'diagnostic': 'diagnostic_name', 'dataset': 'CanESM5', 'grid': 'gn', 'recipe_dataset_index': 1, 'institute': ['CCCma'], 'activity': 'CMIP', 'alias': 'CMIP6', 'original_short_name': 'sftof', 'standard_name': 'sea_area_fraction', 'long_name': 'Sea Area Percentage', 'units': '%', 'modeling_realm': ['ocean'], 'frequency': 'fx', 'filename': '/tmp/pytest-of-valeriu/pytest-96/popen-gw0/test_landmask_no_fx0/input/sftof_Lmon_CanESM5_piControl_r1i1p1f1_gn_1990_1999.nc'}])
--------------------------------------------------------------------- Captured log call ----------------------------------------------------------------------
WARNING  esmvalcore._recipe:_recipe.py:455 Missing data for fx variable 'sftlf'
WARNING  esmvalcore._recipe:_recipe.py:455 Missing data for fx variable 'sftof'
WARNING  esmvalcore._recipe:_recipe.py:441 Missing data for fx variable sftlf and experiment historical; since data is CMIP6, will try experiment piControl, most fx stuff is there.
WARNING  esmvalcore._recipe:_recipe.py:450 Failed to find data for CMIP6 fx variable sftlf with experiment piControl as well.
WARNING  esmvalcore._recipe:_recipe.py:441 Missing data for fx variable sftof and experiment historical; since data is CMIP6, will try experiment piControl, most fx stuff is there.
WARNING  esmvalcore._recipe:_recipe.py:455 Missing data for fx variable 'sftlf'
```
exactly what I want :grin:

Comment (schlunma) [reply tests/integration/test_recipe.py:2208]: Not exactly what I want :smile: : `sftof_Lmon_CanESM5_piControl_r1i1p1f1_gn_1990_1999.nc`. There is no `sftof` in `Lmon`.

The main problem here is that `fx_info['mip']` is set to `Lmon` after the automatic mip search, which causes the data finder to not fail (since the `mip` is not `fx` anymore). In this case you need to repeat the following logic (i.e. the automatic mip search) also for the new `exp` (after removing `mip` again from `fx_info` if necessary):

https://github.com/ESMValGroup/ESMValCore/blob/d75eb9102bf1938bb33da2034c8a25973651532d/esmvalcore/_recipe.py#L418-L433

File: tests/integration/test_recipe.py (hunk 5 above)
Comment (valeriupredoi) [reply tests/integration/test_recipe.py]: OK I fixed this now :beer:

**Comment by zklaus** [comment]
Let's put some numbers to the problem. Looking only at `(Ofx, areacello)` in `CMIP` experiments using the data pool at Jasmin, I find the following statistics for the published fx files. Here, number gives the number of unique (source_id, ensemble) combinations that publish these fx files under the given experiments.

|exp|number|
|-|-|
|historical|23|
|piControl,1pctCO2,historical|14|
|abrupt-4xCO2,historical|13|
|piControl,historical|12|
|piControl|10|
|amip,historical|9|
|piControl,1pctCO2,esm-piControl,historical|3|
|abrupt-4xCO2|3|
|1pctCO2,abrupt-4xCO2,historical|2|
|1pctCO2,historical|2|
|1pctCO2|2|
|amip,piControl,1pctCO2,abrupt-4xCO2,historical|1|
|esm-piControl-spinup,esm-hist,piControl,piControl-spinup,esm-piControl,1pctCO2,abrupt-4xCO2,historical|1|
|piControl,1pctCO2|1|
|piControl,1pctCO2,abrupt-4xCO2,historical|1|
|piControl,amip,1pctCO2,historical|1|
|piControl,piControl-spinup,1pctCO2,abrupt-4xCO2,historical|1|

In other words, `historical` is more prevalent than `piControl`, being absent from only 16 entries, but really there is no clear system. In light of this, I again strongly suggest that we continue the already ongoing discussions about how to approach this instead of moving forward with this PR.

**Review by zklaus — CHANGES_REQUESTED** [review]
The assumption that FX data is generally available in the piControl experiment is not borne out by the data, hence further discussion is required.

**Comment by valeriupredoi** [comment]
wait, I don;t understand the numbers:


historical | 23
-- | --
piControl,1pctCO2,historical | 14

Does this mean that 23 models have (Ofx, areacello) in historical and 14 models have it in piControl,1pctCO2, and historical?

**Comment by zklaus** [comment]
Yes, to be precise 23 (model, ensemble) combinations have the data only in `historical`, 14 (model, ensemble) combinations have it in all three of piControl, 1pctCO2, and historical.

File: tests/integration/test_recipe.py (hunk 4 above)
Comment (valeriupredoi) [reply tests/integration/test_recipe.py:2208]: aye! but why is Lmon added as a valid mip for sftof in the first place? All my new bit of code is doing is jiggling the exp for a set of otherwise fixed and hopefully valid data parameters?

**Comment by valeriupredoi** [comment]
OK but that's not a relevant count since some models have a large number of ensembles, those 23 there might be from a single model, which is probably the IPSL one :grinning: The fact that you see piControl in every row of your table is telling you why this PR is in :beer:

**Comment by valeriupredoi** [comment]
what's interesting from your table is that historical is prob as well populated with fx data as piControl, so we can do a search there too

File: tests/integration/test_recipe.py (hunk 4 above)
Comment (schlunma) [reply tests/integration/test_recipe.py:2208]: That's why:

https://github.com/ESMValGroup/ESMValCore/blob/d75eb9102bf1938bb33da2034c8a25973651532d/esmvalcore/_recipe.py#L389-L391

I think this was necessary since code at a later stage needs the `mip` attribute.

Comment (valeriupredoi) [reply tests/integration/test_recipe.py:2208]: OK but this is not good since we need to set the correct data/file parameters right before we call  `_get_input_files(fx_info, config_user)[0]` - why are we running that function with invalid data descriptors, like sftof-Lmon?

Comment (schlunma) [reply tests/integration/test_recipe.py:2208]: We didn't do that before the changes here :wink: As I said, you need to make sure to use same logic as for the original `exp`: If `mip` exists -> use this; if not -> call `_search_fx_mip`

Comment (valeriupredoi) [reply tests/integration/test_recipe.py:2208]: as in - the moment we call the data finder function we need to make sure that `fx_info` dict is correct and not massage it afterwards if we don't find files, unless one changes the exp to piControl which is a CMOR table-independent parameter (like I'm doing here) - what am I missing here?

Comment (valeriupredoi) [reply tests/integration/test_recipe.py:2208]: arite, that whole block needs to be a function on its own, and an annoying one too, since that needs to be called once (ideally) and the `fx_info` needs to be set correctly before calling the data finder file fetter bleh :grin:

**Comment by zklaus** [comment]
> OK but that's not a relevant count since some models have a large number of ensembles, those 23 there might be from a single model, which is probably the IPSL one.

There might be some of that going on, but manual inspection suggests that is not the main reason. But of course, feel free to suggest a better metric, or better yet, let us know which numbers you used to justify this approach.

> The fact that you see piControl in every row of your table is telling you why this PR is in.

You mean the fact that we **don't** see it in every row and indeed missing in many and some of the most important rows let's us take a step back and re-evaluate the approach, right?

File: tests/integration/test_recipe.py (hunk 4 above)
Comment (valeriupredoi) [reply tests/integration/test_recipe.py]: OK Manu, I have made that mip jiggling a function now, that gets applied before we look for piControl, rehabilitated the old test that needed to stay the same and added two failsafe tests that should fail without using the `patched_datafinder` - these cover the new functionality with piControl, but they cover it synthetically, I don't want to start using yaml files containing actual paths to filesso I can write a test that doesn't find eg historical but finds piControl, that's overkill for this PR, what do you think?

**Comment by valeriupredoi** [comment]
OK @zklaus so what do you suggest right now? @ledm (and possibly other, less vocal, users) are faced with a long-term issue that fx data is in all sorts of dirs where esmvaltool is not looking for, and the wildcard PR is stagnant still - and to be honest with you I'd rather search in a subset of exps than in **all** exps - one, less time consuming and two - God only knows what crazy fx files are in an `abrupt-4xCO2-monkeysOnAPlane` experiment, at least we know piControl is solid data, OK maybe historical too. Anyway, am open to suggestions, but am also keen on a rather sooner than later solution, even if it's a bit kluge-y :beer:

**Comment by schlunma** [comment]
Cheers @valeriupredoi for all the changes :beer: I have some additional comments, but let's continue the review once we've sorted out how to proceed with this!

**Comment by ledm** [comment]
From my side, now that there's the chance to use wildcards (#1082) to find the FX files, I'm not sure that we need to have a default location. (Sorry @valeriupredoi).

**Comment by zklaus** [comment]
@valeriupredoi, I mainly suggest two things:

First, following the standard procedure, that is
1. Start from an actual problem. If you don't immediately have an error to go off, but still suspect a problem, collect the information that shows the problem.
2. Open an issue to discuss the problem and how to address it. Rarely is one person in a position of complete awareness of implications and boundary conditions.
3. If and when reasonable consensus is reached, start working on a draft PR. Pay attention to the checklist; it's there for a reason and we should not think that as core or lead or whatever developers we don't have to follow it, but rather exemplify how and why following it improves the quality.

In this part, the main problem with this PR is that it is lacking in step 1 because it starts from an unsubstantiated claim. Whether that claim turns out to be true or not is almost immaterial, but given the data, I wouldn't call it completely true.

Step 2 was just left out completely.
This is particularly troubling because there is actually a lot of discussion on this precise topic ongoing, see for example https://github.com/ESMValGroup/ESMValCore/issues/1282#issuecomment-919030036, https://github.com/ESMValGroup/ESMValCore/issues/1138#issuecomment-845202897, and ESMValGroup/ESMValTool#2308, with a lot of good insights and ideas.
This PR just sidesteps all of that information and all the aspects covered in the discussion.
Obscuring this sidestepping of the community, the checklist for step 3 was modified to remove the issue requirement.


Ok, so that was the first thing. The second thing is that we need to prioritize. Why do you reckon the wildcard PR (ESMValGroup/ESMValCore#1082) is stagnant? I think it has something to do with us being more eager to start new things than to finish old ones. Don't get me wrong. I think this is very human and I certainly have the same urge. I firmly believe that we need to get better here and using milestones to highlight imminent work is one tool for doing this. Of course, there are other tools, and we can always discuss how to do this, but any way you slice it, we need to close old issues and PRs, preferably by solving and finishing them.

**Comment by valeriupredoi** [comment]
@schlunma cheers! @ledm all good by me, as long as you think you can proceed with work @zklaus nice pep talk and analysis, I am aware of all the procedures, but sometimes if you stick too much to procedures you end up with a procedurally sound, half-working machinery - I am more of a dirty fix before we fix it proper kind of guy :grin: So consensus is to close this and go on with #1082 ? Are there any known risks of searching all the experiments - my only concern is that some crazy experiment might have some crazy fx data that's non-standard - if this is not the case, then we can defo forget about this PR :+1:

**Comment by zklaus** [comment]
Well, in this case, it helped us avoid unnecessary changes. Unfortunately only after wasting a lot of resources.
